frontend/src/test/java/com/google/tripmeout/frontend/servlet/PlaceParentServletTest.java

```diff
@@ -0,0 +1,222 @@
+                                                .setName("CA Trip")
+                                                .setTripId("abc123")
+                                                .setPlaceId("CA, USA")
+                                                .setLatitude(111.111)
+                                                .setLongitude(45.4)
+                                                .setUserMark(PlaceVisitModel.UserMark.YES)
+                                                .build();
+
+  private static final PlaceVisitModel UK_B = PlaceVisitModel.builder()
+                                                  .setName("UK Trip")
+                                                  .setTripId("123abc")
+                                                  .setPlaceId("ChIJ3S-JXmauEmsRUcIaWtf4MzE")
+                                                  .setLatitude(111.111)
+                                                  .setLongitude(45.4)
+                                                  .setUserMark(PlaceVisitModel.UserMark.MAYBE)
+                                                  .build();
+
+  @Before
+  public void setUp() {
+    initMocks(this);
+    Injector injector = Guice.createInjector(new GsonModelSerializationModule());
+    this.gson = injector.getInstance(Gson.class);
+  }
+
+  @Test
+  public void doPost_badUri_setsBadRequestStatus() throws IOException {
+    FakeHttpServletResponse response = new FakeHttpServletResponse();
+    PlaceVisitStorage placeStorage = new InMemoryPlaceVisitStorage();
+    PlaceParentServlet servlet = new PlaceParentServlet(placeStorage, gson);
+    when(request.getRequestURI()).thenReturn("/trips-number/abc123/placeVisits");
+    when(request.getReader())
+        .thenReturn(new BufferedReader(
+            new StringReader("{tripId: abc123, placeId: ChIJ3S-JXmauEmsRUcIaWtf4MzE}")));
+
+    servlet.doPost(request, response);
+
+    assertThat(response.getStatus()).isEqualTo(HttpServletResponse.SC_BAD_REQUEST);
+  }
+
+  @Test
+  public void doPost_badPlaceId_setsNotFoundStatus() throws IOException {
+    FakeHttpServletResponse response = new FakeHttpServletResponse();
+    PlaceVisitStorage placeStorage = new InMemoryPlaceVisitStorage();
+    PlaceParentServlet servlet = new PlaceParentServlet(placeStorage, gson);
+    when(request.getRequestURI()).thenReturn("/trips/abc123/placeVisits");
+    when(request.getReader())
+        .thenReturn(new BufferedReader(new StringReader("{tripId: abc123, placeId: hello}")));
+
+    servlet.doPost(request, response);
+    Optional<PlaceVisitModel> place = placeStorage.getPlaceVisit("abc123", "hello");
+
+    assertThat(place).isEmpty();
+    assertThat(response.getStatus()).isEqualTo(HttpServletResponse.SC_NOT_FOUND);
+  }
+
+  @Test
+  public void doPost_goodRequest_storesPlaceVisitModel()
+      throws IOException, ApiException, InterruptedException {
+    PlaceDetailsRequest placeRequest = new PlaceDetailsRequest(CONTEXT)
```

Comment:
Our unit tests should _not_ be querying real external services like this. Instead, we should introduce a place service interface (described in #64) and mock that where needed.

In this particular case, there's no need to even call a fake service--we only want the resulting `PlaceDetails` for comparison with the result. Using a real, external service such as this also introduces test flakiness for various reasons:
- The service could go down during the test.
- Place details may change between the two calls to the real service (here, where we're computing the expected result and within the servlet, where we query the service again).
- The test could otherwise fail out or time out while waiting for a response from the service.